proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,682 @@
+  [N.B. In the easy case when we don't change the ABI, compiler bootstrapping stages and TH stages coincide!
+  Stage 2 - 1 = Stage 1.]
+
+  Switching all existing ``derive-*`` code generators to TH would probably make them lighter and easier to maintain.
+  It should also allow building GHC the binary with plain `cabal`.
+  Hadrian would be one step closer to being another implementation of Cabal/cabal-install without GHC-specific logic.
+
+``*_BUILD_*`` and ``*_HOST_*`` not always defined
+  The conditional definition of the CPP macros ensures they don't pollute the purity of the build when they don't matter.
+  This is important for highly pure build systems like Nix to not have to needless rebuild stuff when the target platform changes.
+  It will also cut down on people improperly using "target" when they meant "host".
+
+``Lift`` and qualified goals
+  When we carefully introduce lifting to stage-hygienic goals, we need to ensure that the type being lifted is the same or "close enough".
+  This means we will need to introduce a intra-package constraint on package defining that type across the stage pair where Lift is made available.
+  Conceptually, there might be an auto-generated package with the orphan ``Lift`` instance which imposes the same version constraint on it's library dependency in both stages.
+
+  In particular, existing qualified dependencies from ``setup-depends`` and ``build-tool-depends`` are from stage *n* to *n - 1*;
+  that the stages are different alone explains why versions are allowed to differ.
+  In particular this means given a dependency edge where the needed and needing components are in the same package regardless of their relative stage indices,
+  the same version of the package must be used for both.
+
+Static and dynamic linking
+  It's hard to build large swaths of the ecosystem with just static libraries. [#thanks-mboes]_
+  This is because GHCi "prefers" loading dynamic libraries.
+  The easiest thing to do is build both ways and then throw away the shared libraries.
+  The better thing to do is think of shared GHC -> static outputs as cross compilation.
+  Then only stage < 0 code for splices needs be built as shared libraries (or bytecode).
+  Plus, that builds aren't installed along side the stage 0 code, so there is no extra step to manually shave off the shared libraries.
+
+Costs and Drawbacks
+-------------------
+
+- This is a huge amount of work.
+  But I am fine chipping away it over a long period of time.
+
+- Even a temporary conflict between typed TH and this could slow typed TH's adoption.
+
+- I don't know of precedent for extensions that prevent modules from being linked together.
+
+- Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same Cabal component but in a different module.
```

Comment:
```suggestion
- This is a huge amount of work.
  But I am fine chipping it away over a long period of time.

- Even a temporary conflict between typed TH and this could slow down typed TH's adoption.

- I don't know of a precedent for extensions that prevent modules from being linked together.

- Most existing libraries with commonly used TH helpers (``lens``, ``aeson``) have the TH in the same Cabal component but in a different module.
```